### PR TITLE
Auto uppercase Gcode except commands in configurable blacklist

### PR DESCRIPTION
**Pull request by agarwali**
Thank you for your interest into contributing to OctoPrint, it's
highly appreciated!

Please make sure you have read the "guidelines for contributing" as
linked just above this form, there's a section on Pull Requests in there
as well that contains important information.

As a summary, please make sure you have ticked all points on this
checklist:
- [x] Your changes are not possible to do through a plugin and relevant
  to a large audience (ideally all users of OctoPrint)
- [x] If your changes are large or otherwise disruptive: You have
  made sure your changes don't interfere with current development by
  talking it through with the maintainers, e.g. through a
  Brainstorming ticket
- [x] Your PR targets OctoPrint's devel branch (not master,
  maintenance or anything else)
- [x] Your PR was opened from a custom branch on your repository
  (no PRs from your version of master, maintenance or devel please),
  e.g. dev/my_new_feature
- [x] Your PR only contains relevant changes: no unrelated files,
  no dead code, ideally only one commit - rebase your PR if necessary!
- [x] Your changes follow the coding style
- [x] If your changes include style sheets: You have modified the
  .less source files, not the .css files (those are generated with
  lessc)
- [x] You have tested your changes (please state how!) - ideally you
  have added unit tests
- [x] You have run the existing unit tests against your changes and
  nothing broke
- [x] You have added yourself to the AUTHORS.md file :)

Feel free to delete all this help text, then describe
your PR further. You may use the template provided below to do that.
The more details the better!

---
#### What does this PR do and why is it necessary?

This PR enables auto upper casing of the G-Code commands. A new config setting was added: auto uppercase blacklist that is configurable in the advanced option of the serial tab, with M117 included in the default settings of the blacklist. The sendCommand method in the TerminalViewModel is changed to auto uppercase the entire G-code command except for commands in blacklist.
Updated documentation to include the default auto uppercase blacklist command M117.
#### How was it tested? How can it be tested by the reviewer?

This was tested on a 3D printer with all possible cases of G-code commands that were checked in the G-code terminal i) Commands in the blacklist, and ii) commands not in the blacklist.
#### What are the relevant tickets if any?

Resolves: #1026 
#### Further notes

Resolves: #1026

### CODE REVIEW

**Comment by wolfmanjm** [comment]
please note this will break the ability to send console commands to Smoothie which uses lowercase for the commands and uppercase for G/M codes. We increasingly have more useful console commands in smoothie and one of the nice thigns about octoprint is it sends them without needing any prefixes (like Pronterface requires the @ prefix).

**Comment by foosel** [comment]
@wolfmanjm unless I misunderstand you, it shouldn't be a problem since it will only uppercase detected gcode (g, m or  the followed by a number) that's not blacklisted. So if you send "g0 x10" it will turn it into "G0 X10", but it will not touch "hello world" or any other arbitrary string that doesn't match `^[GMTgmt]\d+.*`

File: src/octoprint/static/js/app/viewmodels/terminal.js [hunk 1]
```diff
@@ -259,10 +259,14 @@ $(function() {
...
             var commandMatch = command.match(re);
+            self.blacklist =  splitTextToArray(self.settings.serial_autoUppercaseBlacklist(), ",", true);
             if (commandMatch != null) {
                 command = commandMatch[1].toUpperCase() + ((commandMatch[2] !== undefined) ? commandMatch[2] : "");
+                if (self.blacklist.indexOf(commandMatch[1].toUpperCase()) < 0){
+                    command = command.toUpperCase()
+                }
             }
-
+
```
Comment (BillyBlaze) [inline src/octoprint/static/js/app/viewmodels/terminal.js]: Can you fix this leading whitespace? A good editior should remove it by default if you saved (e.g. Atom)

File: src/octoprint/util/comm.py [hunk 2]
```diff
@@ -566,7 +567,7 @@ def sendCommand(self, cmd, cmd_type=None, processed=False, force=False):
 			cmd = process_gcode_line(cmd)
 			if not cmd:
 				return
-
+
```
Comment (BillyBlaze) [inline src/octoprint/util/comm.py]: Ditto

**Comment by agarwali** [comment]
@wolfmanjm, @foosel is right, it will not auto uppercase any arbitrary string unless the first part of the string matches /^([gmt][0-9]+)(\s.*)?/. If any smoothie command happens to start with [gmt][0-9], that we do not want to auto uppercase, it needs to be added to the blacklist.

**Comment by wolfmanjm** [comment]
Oh excellent thanks. I read it to mean it upper cased everything unless it was blacklisted :)
So things like get pos will work and stay lower case :)

File: src/octoprint/static/js/app/viewmodels/terminal.js (hunk 1 above)
Comment (agarwali) [inline src/octoprint/static/js/app/viewmodels/terminal.js]: @BillyBlaze, it is now taken care of

File: src/octoprint/static/js/app/viewmodels/terminal.js [hunk 3]
```diff
@@ -259,8 +259,12 @@ $(function() {
 
             var re = /^([gmt][0-9]+)(\s.*)?/;
             var commandMatch = command.match(re);
+            self.blacklist =  splitTextToArray(self.settings.serial_autoUppercaseBlacklist(), ",", true);
```
Comment (foosel) [inline src/octoprint/static/js/app/viewmodels/terminal.js]: Setting an (undeclared) instance variable like this within a method is unclean. Especially since you don't use this instance variable outside of this method.

Two ways to go about this. One is simply declaring a local `var blacklist` instead of `self.blacklist`. That has the disadvantage however that on ever single command that is send the whole blacklist will be fetched from the settings view model and turned into an array before matching.

Cleaner would be to use `self.settings.serial_autoUppercaseBlacklist.subscribe(...)` to be notified of changes to the value (see [Kockout's docs for details on subscribe](http://knockoutjs.com/documentation/observables.html#explicitly-subscribing-to-observables)) and then updating a (previously declared) `self.blacklist` with the array-ified value that in turn is then used in the `sendCommand` routine.

So basically: 

  * add `self.blacklist` to the view model properly (top area), initialize to empty array
  * subscribe to `self.settings.serial_autoUppercaseBlacklist`:

    ``` javascript
    self.settings.serial_autoUppercaseBlacklist(function(newValue) {
        self.blacklist = splitTextToArray(newValue, ",", true);
    });
    ```
  * Use `self.blacklist` in `sendCommand`

File: src/octoprint/util/comm.py [hunk 4]
```diff
@@ -270,6 +270,7 @@ def __init__(self, port = None, baudrate=None, callbackObject=None, printerProfi
 
 		self._long_running_commands = settings().get(["serial", "longRunningCommands"])
 		self._checksum_requiring_commands = settings().get(["serial", "checksumRequiringCommands"])
+		self._auto_uppercase_blacklist = settings().get(["serial", "autoUppercaseBlacklist"])
```
Comment (foosel) [inline src/octoprint/util/comm.py]: What is this needed for?

**Comment by foosel** [comment]
Hey, took a look at the PR and found a couple of points, could you take a look at the comments? Thanks :)

**Comment by foosel** [comment]
Ok, this has now been rebased against `devel` and then cherry picked as 51406b7 and ce16f86 to `maintenance`, as part of #2177. Thanks to @pbackx for picking up the reigns and taking care of the review comments 👍 

Closing manually since Github doesn't pick up on cherry-picks.
